Traceback (most recent call last):
  File "/home/runner/micromamba/envs/comp2421-book/lib/python3.12/site-packages/jupyter_cache/executors/utils.py", line 58, in single_nb_execution
    executenb(
  File "/home/runner/micromamba/envs/comp2421-book/lib/python3.12/site-packages/nbclient/client.py", line 1305, in execute
    return NotebookClient(nb=nb, resources=resources, km=km, **kwargs).execute()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/micromamba/envs/comp2421-book/lib/python3.12/site-packages/jupyter_core/utils/__init__.py", line 165, in wrapped
    return loop.run_until_complete(inner)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/micromamba/envs/comp2421-book/lib/python3.12/asyncio/base_events.py", line 687, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "/home/runner/micromamba/envs/comp2421-book/lib/python3.12/site-packages/nbclient/client.py", line 705, in async_execute
    await self.async_execute_cell(
  File "/home/runner/micromamba/envs/comp2421-book/lib/python3.12/site-packages/nbclient/client.py", line 1058, in async_execute_cell
    await self._check_raise_for_error(cell, cell_index, exec_reply)
  File "/home/runner/micromamba/envs/comp2421-book/lib/python3.12/site-packages/nbclient/client.py", line 914, in _check_raise_for_error
    raise CellExecutionError.from_cell_and_msg(cell, exec_reply_content)
nbclient.exceptions.CellExecutionError: An error occurred while executing the following cell:
------------------
def f(x):
    return x**3


def df(x):
    return 3 * x**2


x = 1

headers = ["dx", "approx", "abs error", "rel error"]
data = []

for e in range(4, 18, 2):
    dx = 10**-e
    approx = (f(x + dx) - f(x)) / dx
    exact = df(x)
    abs_error = abs(exact - approx)
    rel_error = abs(exact - approx) / exact

    new_data = [dx, approx, abs_error, rel_error]
    data.append(new_data)

import pandas as pd

df = pd.DataFrame(data, columns=headers)
df.style.format(
    formatter={"dx": "{:e}", "approx": "{:f}", "abs error": "{:e}", "rel error": "{:e}"}
).hide_index().set_caption(
    "Simple approximation of a derivative using floating point arithmetic"
)
------------------


[0;31m---------------------------------------------------------------------------[0m
[0;31mAttributeError[0m                            Traceback (most recent call last)
Cell [0;32mIn[1], line 29[0m
[1;32m     24[0m [38;5;28;01mimport[39;00m [38;5;21;01mpandas[39;00m [38;5;28;01mas[39;00m [38;5;21;01mpd[39;00m
[1;32m     26[0m df [38;5;241m=[39m pd[38;5;241m.[39mDataFrame(data, columns[38;5;241m=[39mheaders)
[1;32m     27[0m [43mdf[49m[38;5;241;43m.[39;49m[43mstyle[49m[38;5;241;43m.[39;49m[43mformat[49m[43m([49m
[1;32m     28[0m [43m    [49m[43mformatter[49m[38;5;241;43m=[39;49m[43m{[49m[38;5;124;43m"[39;49m[38;5;124;43mdx[39;49m[38;5;124;43m"[39;49m[43m:[49m[43m [49m[38;5;124;43m"[39;49m[38;5;132;43;01m{:e}[39;49;00m[38;5;124;43m"[39;49m[43m,[49m[43m [49m[38;5;124;43m"[39;49m[38;5;124;43mapprox[39;49m[38;5;124;43m"[39;49m[43m:[49m[43m [49m[38;5;124;43m"[39;49m[38;5;132;43;01m{:f}[39;49;00m[38;5;124;43m"[39;49m[43m,[49m[43m [49m[38;5;124;43m"[39;49m[38;5;124;43mabs error[39;49m[38;5;124;43m"[39;49m[43m:[49m[43m [49m[38;5;124;43m"[39;49m[38;5;132;43;01m{:e}[39;49;00m[38;5;124;43m"[39;49m[43m,[49m[43m [49m[38;5;124;43m"[39;49m[38;5;124;43mrel error[39;49m[38;5;124;43m"[39;49m[43m:[49m[43m [49m[38;5;124;43m"[39;49m[38;5;132;43;01m{:e}[39;49;00m[38;5;124;43m"[39;49m[43m}[49m
[0;32m---> 29[0m [43m)[49m[38;5;241;43m.[39;49m[43mhide_index[49m()[38;5;241m.[39mset_caption(
[1;32m     30[0m     [38;5;124m"[39m[38;5;124mSimple approximation of a derivative using floating point arithmetic[39m[38;5;124m"[39m
[1;32m     31[0m )

[0;31mAttributeError[0m: 'Styler' object has no attribute 'hide_index'

